cranelift/codegen/src/ir/pcc.rs

```diff
@@ -394,6 +395,42 @@ impl Expr {
             None
         }
     }
+
+    /// Multiply an expression by a constant, if possible.
+    fn scale(&self, factor: u32) -> Option<Expr> {
+        let offset = self.offset.checked_mul(i128::from(factor))?;
+        match self.base {
+            BaseExpr::None => Some(Expr {
+                base: BaseExpr::None,
+                offset,
+            }),
+            BaseExpr::Max => Some(Expr {
+                base: BaseExpr::Max,
+                offset: 0,
+            }),
+            _ => None,
+        }
+    }
+
+    /// Multiply an expression by a constant, rounding downward if we
+    /// must approximate.
+    fn scale_downward(&self, factor: u32) -> Expr {
+        self.scale(factor).unwrap_or(Expr::constant(0))
+    }
```

Comment:
This feels like a suspicious operation. Rounding down seems sus when we are talking about proving that accesses are in-bounds!

Reply:
Indeed! It's specifically for lower bounds: when scaling up, we want lower bounds to approximate *downward*, not upward. I'll add that use-case to the comment.

Reply:
Added some comments regarding this to help clarify.